README.md
# PE Tree
PE Tree is a Python module for viewing [Portable Executable (PE) files](https://en.wikipedia.org/wiki/Portable_Executable) in a tree-view using [pefile](https://github.com/erocarrera/pefile) and [PyQt5](https://pypi.org/project/PyQt5/).
It can also be used with [IDA Pro](https://www.hex-rays.com/products/ida/), [Ghidra](https://ghidra-sre.org/), [Volatility](https://github.com/volatilityfoundation/volatility3), [Rekall](http://www.rekall-forensic.com/) and [minidump](https://github.com/skelsec/minidump) to view and dump in-memory PE files, as well as perform import table reconstruction.

## Table of contents

1. [Features](#features)
2. [Application](#application)
    * [Requirements](#requirements)
    * [Features](#features-1)
    * [Installation](#installation)
        * [Windows](#windows)
        * [Mac/Linux](#maclinux)
        * [For developers](#for-developers)
    * [Usage](#usage)
    * [Dark-mode](#dark-mode)
3. [IDAPython](#idapython)
    * [Requirements](#requirements-1)
    * [Features](#features-2)
    * [Installation](#installation-1)
        * [Using setuptools](#using-setuptools)
        * [Install manually](#install-manually)
        * [For developers](#for-developers-1)
    * [Usage](#usage)
    * [Example](#example)
        * [Dumping in-memory PE files](#dumping-in-memory-pe-files)
4. [Rekall](#rekall)
    * [Requirements](#requirements-2)
    * [Features](#features-3)
    * [Installation](#installation-2)
    * [Usage](#usage-2)
5. [Volatility](#volatility)
    * [Requirements](#requirements-3)
    * [Features](#features-4)
    * [Installation](#installation-3)
    * [Usage](#usage-3)
6. [Ghidra](#ghidra)
    * [Requirements](#requirements-4)
    * [Features](#features-5)
    * [Installation](#installation-4)
    * [Usage](#usage-4)
7. [Minidump](#minidump)
    * [Requirements](#requirements-5)
    * [Features](#features-6)
    * [Installation](#installation-5)
    * [Usage](#usage-5)
8. [Configuration](#configuration)
    * [Overview](#overview)
    * [Options](#options)
    * [Location](#location)
    * [3rd party data sharing](#3rd-party-data-sharing)
9. [Troubleshooting](#troubleshooting)
10. [Contributing](#contributing)
    * [Developer documentation](#developer-documentation)
11. [License](#license)

## Features
* Standalone application with plugins for:
    * IDA Pro
    * Ghidra
    * Volatility
    * Rekall
    * Minidumps
    * Carving
* Supports Windows, Linux and Mac
* Parsing PE files and memory images from:
    * File-system
    * ZIP archives (including password protected)
    * Windows memory dumps (raw, EWF, vmem etc.)
    * Live Windows memory (using rekall)
    * Windows Minidump
    * IDA Pro database
    * Ghidra database
    * Binary file carving
* Rainbow PE map:
    * Provides a high-level overview of PE structures, size and file location
    * Allows for fast visual overview and comparison of PE samples
* Displays the following PE headers in a tree-view:
    * MZ header
    * DOS stub
    * Rich headers
    * NT/File/Optional headers
    * Data directories
    * Sections
    * Imports
    * Exports
    * Debug information
    * Load config
    * TLS
    * Resources
    * Version information
    * Certificates
    * Overlay
* Extract and save data from:
    * DOS stub
    * Sections
    * Resources
    * Certificates
    * Overlay
    * Export to CyberChef for further manipulation
* Perform VirusTotal searches of:
    * File hashes
    * PDB path
    * Timestamps
    * Section hash/name
    * Import hash/name
    * Export name
    * Resource hash
    * Certificate serial
* Dump loaded PE images from memory:
    * Fix up section pointers and sizes
    * Fix up PE headers:
        * Remove unnecessary data directory pointers
        * Recalculate PE checksum
        * Update entry-point
    * Reconstruct import address and directory tables (IAT/IDT) using several methods:
        1. Use existing IAT/IDT
        2. Rebuild IDT from existing IAT
        3. Rebuild IAT and IDT from disassembly (using IDA Pro, Ghidra or capstone)

## Application

>*The PE Tree standalone application finds portable executables in files, folders and ZIP archives.*

![PE Tree standalone application](./doc/images/pe_tree.png)

### Requirements
* Python 3.5+

### Features
* Scan files and folders for PE files
* Extract PE files from ZIP archives (including password protected with *infected*)
* Carve PE files from binary files
* Double-click VA/RVA to disassemble with capstone
* Hex-dump data

### Installation

#### Using pip (recommended)
Install directly from GitHub using a fresh virtual environment and pip:

##### Windows
```
> virtualenv env
> env\Scripts\activate
> pip install --upgrade pip
> pip install git+https://github.com/blackberry/pe_tree.git
```

##### Mac/Linux
```
$ python3 -m venv env
$ source ./env/bin/activate
$ pip install --upgrade pip
$ pip install git+https://github.com/blackberry/pe_tree.git
```
#### For developers
Git clone the repository and setup for development:

##### Windows
```
> git clone https://github.com/blackberry/pe_tree.git
> cd pe_tree
> virtualenv env
> env\Scripts\activate
> pip install -e .
```

##### Mac/Linux
```
$ git clone https://github.com/blackberry/pe_tree.git
$ cd pe_tree
$ python3 -m venv env
$ source ./env/bin/activate
$ pip install -e .
```

### Usage

Run PE Tree and scan for portable executables in files, folders and ZIP archives:

```
$ pe-tree -h
usage: pe-tree [-h] [filenames [filenames ...]]

PE-Tree

positional arguments:
  filenames   Path(s) to file/folder/zip

optional arguments:
  -h, --help  show this help message and exit
```

Run PE Tree and attempt to [carve](https://en.wikipedia.org/wiki/File_carving) portable executable files from a binary file:

```
$ pe-tree-carve -h
usage: pe-tree-carve [-h] filename

PE-Tree (Carve)

positional arguments:
  filename    Path to file to carve

optional arguments:
  -h, --help  show this help message and exit
```

#### Dark-mode
Dark-mode can be enabled by installing [QDarkStyle](https://pypi.org/project/QDarkStyle/):
```
$ pip install qdarkstyle
```

## IDAPython

>*The PE Tree IDAPython plugin finds portable executables in IDA databases.*

![PE Tree IDAPython plugin](./doc/images/pe_tree_ida.png)

### Requirements
* IDA Pro 7.0+ with Python 2.7
* IDA Pro 7.4+ with Python 2.7 or 3.5+

### Features
* Easy navigation of PE file structures
* Double-click on a memory address in PE Tree to view in *IDA-view* or *hex-view*
* Search an IDB for in-memory PE images and;
    * Reconstruct imports (IAT + IDT)
    * Dump reconstructed PE files
    * Automatically comment PE file structures in IDB
    * Automatically label IAT offsets in IDB

### Installation
To install and run as an IDAPython plugin you can either use setuptools or install manually.

#### Using setuptools
1. Download pe_tree and install for the **global** Python interpreter used by IDA:
   
    ```
    $ git clone https://github.com/blackberry/pe_tree.git
    $ cd pe_tree
    $ python setup.py develop --ida
    ```
3. Copy [pe_tree_ida.py](./pe_tree_ida.py) to your [IDA plugins folder](#ida-plugins-folder)

#### Install manually
1. Download pe_tree and install requirements for the **global** Python interpreter used by IDA:
   
    ```
    $ git clone https://github.com/blackberry/pe_tree.git
    $ cd pe_tree
    $ pip install -r requirements.txt
    ```
3. Copy [pe_tree_ida.py](./pe_tree_ida.py) and the contents of `./pe_tree/` to your [IDA plugins folder](#ida-plugins-folder)

#### For developers
To forgo installing as a plugin, and simply run as a script under IDA, first install the pe_tree package requirements for the **global** Python installation:

```
$ pip install -r requirements.txt
```

Then run [pe_tree_ida.py](./pe_tree_ida.py) under IDA:

`File -> Script file... -> pe_tree_ida.py -> Open`

### IDA plugins folder

| OS      | Plugins folder                              |
| :------ | :------------------------------------------ |
| Windows | `%ProgramFiles%\IDA Pro 7.X\plugins`        |
| Linux   | `/opt/ida-7.X/plugins`                      |
| Mac     | `~/.idapro/plugins`                         |

### Usage

1. Launch IDA Pro and disassemble a PE file (always select **Manual Load** and **Load Resources** for best results!)
2. Load the PE Tree plugin:

    ```Edit -> Plugins -> PE Tree```

### Example

#### Dumping in-memory PE files

Below are the basic steps to dump a packed PE file (for example MPRESS or [UPX](https://upx.github.io/)) and reconstruct imports (assuming the image base/entry-point is fairly standard):

1. Launch IDA Pro and disassemble an MPRESS or UPX packed PE file (select **Manual Load** and **Load Resources**)

2. Select a debugger (Windows or [Bochs](http://bochs.sourceforge.net/)) and run until OEP (usually 0x00401000, but not always!)

3. At this point you could take a memory snapshot (saving all segments) and save the IDB for later

4. Ensure IDA has found all code:

    ```Options -> General -> Analysis -> Reanalyze program```

5. Open the PE Tree IDAPython plugin, right-click in the right-hand pane and select:

    ```Add PE -> Search IDB```

    This will scan the IDB for MZ/PE headers and display any modules it finds.

6. Right-click on **HEADER-0x00400000** (or appropriate module name) and select **Dump...**

7. Specify the **AddressOfEntryPoint** (typically 0x1000, but again, not always!)

8. Ensure **Rebuild IDT/IAT** is selected

9. Dump!

A new executable will be created using the unpacked section data obtained from memory/IDB, whilst a new section named `.pe_tree` and containing the rebuilt IAT, hint name table and IDT will be appended to the PE file (much like an [.idata](https://docs.microsoft.com/en-us/windows/win32/debug/pe-format#the-idata-section) section). 
If the entry-point memory segment has been marked writable during execution (via [VirtualProtect](https://docs.microsoft.com/en-us/windows/win32/api/memoryapi/nf-memoryapi-virtualprotect) for example) then the entry-point section's characteristics will also be marked writable.
Finally, the BASERELOC, BOUND_IMPORT and SECURITY data directories are marked null, and the OPTIONAL_HEADER checksum is recalculated (if specified).

Using the above approach it is possible to dump many in-memory PE files that have either been unpacked, injected, reflectively loaded or hollowed etc.

## Rekall

>*The PE Tree Rekall plugin finds portable executables in Windows memory dumps.*

![PE Tree Rekall plugin](./doc/images/pe_tree_rekall.png)

### Requirements
* Python 3+

### Features
* Operates against a Windows memory dump or a live system 
* View, dump and reconstruct PE modules from;
    * Active processes and DLLs
    * Loaded kernel-mode drivers

### Installation

1. Install Rekall from [GitHub](https://github.com/google/rekall#quick-start).
2. Install PE Tree standalone application (see [Installation](#installation)) under the same virtual environment.

### Usage

Run Rekall and view active processes, DLLs and drivers on a live system:

```
$ rekall --live Memory
[1] Live (Memory) 00:00:00> run -i pe_tree_rekall.py
```

Alternatively, run Rekall/PE Tree against an existing memory dump:

```
$rekall -f memory.vmem
[1] memory.vmem 00:00:00> run -i pe_tree_rekall.py
```

## Volatility

>*The PE Tree Volatility plugin finds portable executables in Windows memory dumps.*

![PE Tree Volatility plugin](./doc/images/pe_tree_volatility.png)

### Requirements
* Python 3.5+

### Features
* Operates against a Windows memory dump
* View, dump and reconstruct PE modules from;
    * Active processes and DLLs
    * Loaded kernel-mode drivers

### Installation

1. Install Volatility3 from [GitHub](https://github.com/volatilityfoundation/volatility3#quick-start).
2. Install PE Tree standalone application (see [Installation](#installation)) under the same virtual environment.

### Usage

```
$ pe-tree-vol -h
usage: pe-tree-vol [-h] filename

PE-Tree (Volatility)

positional arguments:
  filename    Path to memory dump

optional arguments:
  -h, --help  show this help message and exit
```

## Ghidra

>*The PE Tree Ghidra plugin finds portable executables in Ghidra databases.*

![PE Tree Ghidra plugin](./doc/images/pe_tree_ghidra.png)

### Requirements
* Python 3.5+
* Ghidra Bridge

### Features
* Easy navigation of PE file structures
* Double-click on a memory address in PE Tree to view in Ghidra disassembly/hex-view
* Reconstruct imports (IAT + IDT)
* Dump reconstructed PE files

### Installation

1. Install PE Tree (see [Installation](#installation))
2. Install [Ghidra Bridge](https://github.com/justfoxing/ghidra_bridge#install-the-ghidra-bridge-package-and-server-scripts)

### Usage

1. Start the [Ghidra Bridge server](https://github.com/justfoxing/ghidra_bridge#start-server)
2. Run the PE Tree Ghidra plugin

```
$ pe-tree-ghidra -h
usage: pe-tree-ghidra [-h] [--server SERVER] [--port PORT]

PE-Tree (Ghidra)

optional arguments:
  -h, --help       show this help message and exit
  --server SERVER  Ghidra bridge server IP (default: 127.0.0.1)
  --port PORT      Ghidra bridge server port (default: 4768)
```

## Minidump

>*The PE Tree Minidump plugin finds portable executables in Windows Minidumps.*

![PE Tree Minidump plugin](./doc/images/pe_tree_minidump.png)

### Requirements
* Python 3.6+
* [minidump](https://github.com/skelsec/minidump)

### Features
* View, dump and reconstruct PE modules from a Windows Minidump (.dmp) file

### Installation

1. Install PE Tree (see [Installation](#installation))
2. Install [minidump](https://github.com/skelsec/minidump) (`pip install minidump`)

### Usage

```
$ pe-tree-minidump -h
usage: pe-tree-minidump [-h] filename

PE-Tree (Minidump)

positional arguments:
  filename    Path to .dmp file

optional arguments:
  -h, --help  show this help message and exit
```

## Configuration

### Overview

The configuration is stored in an INI file and defaults to the following values:

```
[config]
debug = False
fonts = Consolas,Monospace,Courier
passwords = ,infected
virustotal_url = https://www.virustotal.com/gui/search
cyberchef_url = https://gchq.github.io/CyberChef
```

### Options

| Section | Option                  | Type    | Description                                              |
| :------ | :---------------------- | :------ | :------------------------------------------------------- |
| config  | debug                   | boolean | Print pefile.dump() to output                            |
| config  | fonts                   | string  | Comma-separated list of font names for UI                |
| config  | passwords               | string  | Comma-separated list of ZIP file passwords               |
| config  | virustotal_url          | string  | VirusTotal search URL                                    |
| config  | cyberchef_url           | string  | CyberChef URL                                            |

### Location

| Type        | OS        | Path                                    |
|-------------|-----------|:----------------------------------------|
| Application | Windows   | `%TEMP%\pe_tree.ini`                    |
| Application | Linux/Mac | `/tmp/pe_tree.ini`                      |
| IDAPython   | Windows   | `%APPDATA%\HexRays\IDA Pro\pe_tree.ini` |
| IDAPython   | Linux/Mac | `~/.idapro/pe_tree.ini`                 |
| Rekall      | Windows   | `%TEMP%\pe_tree_rekall.ini`             |
| Rekall      | Linux/Mac | `/tmp/pe_tree_rekall.ini`               |
| Volatility  | Windows   | `%TEMP%\pe_tree_volatility.ini`         |
| Volatility  | Linux/Mac | `/tmp/pe_tree_volatility.ini`           |
| Ghidra      | Windows   | `%TEMP%\pe_tree_ghidra.ini`             |
| Ghidra      | Linux/Mac | `/tmp/pe_tree_ghidra.ini`               |
| Minidump    | Windows   | `%TEMP%\pe_tree_minidump.ini`           |
| Minidump    | Linux/Mac | `/tmp/pe_tree_minidump.ini`             |
| Carve       | Windows   | `%TEMP%\pe_tree_carve.ini`              |
| Carve       | Linux/Mac | `/tmp/pe_tree_carve.ini`                |

### 3rd party data sharing

The following information will be shared with 3rd party web-applications (depending on configuration) under the following conditions:

#### VirusTotal

If the [VirusTotal](https://www.virustotal.com/) URL is specified in the configuration then metadata such as file hashes, timestamps, etc will be sent to VirusTotal for processing when the user clicks on highlighted links or selects "VirusTotal search" from the right-click context menu.

#### CyberChef

If the [CyberChef](https://github.com/gchq/CyberChef) URL is present in the configuration then any file data will be base64 encoded and sent to CyberChef for processing when the user selects "CyberChef" from the right-click context menu.

## Troubleshooting

#### AttributeError: module 'pip' has no attribute 'main'
or

#### PyQt5 fails to install under Linux
Try to upgrade pip to version 20.0+:
```
$ pip install --upgrade pip
```

#### ModuleNotFoundError: No module named 'PyQt5.sip'
Try uninstalling and reinstalling PyQt5 as follows:
```
pip uninstall PyQt5
pip uninstall PyQt5-sip
pip install PyQt5 PyQt5-sip
```

#### Missing imports after dumping
Ensure IDA has found and disassembled all code:

```Options -> General -> Analysis -> Reanalyze program```

After this is completed try to dump/rebuild imports again.

## Contributing

Please feel free to contribute! Issues and pull requests are most welcome.

### Developer documentation

To build documentation from source using [Sphinx](https://www.sphinx-doc.org/en/master/):

```
$ pip install sphinx
$ sphinx-apidoc -o ./doc/source/ .
$ sphinx-build -b html ./doc/source ./doc/build -E
```

To view the documentation open `./doc/build/index.html` in a web-browser.

## Authors

*Tom Bonner* - [tombonner](https://github.com/tombonner) - [@thomas_bonner](https://twitter.com/thomas_bonner)

## License

PE Tree is distributed under the Apache License. See [LICENSE](LICENSE) for more information. 


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

## [1.0.30](/../../releases/tag/1.0.30) - 2021-04-08

### Added

- Plugin for [Ghidra](https://ghidra-sre.org/) using [Ghidra Bridge](https://github.com/justfoxing/ghidra_bridge)
- Plugin for [Volatility](https://github.com/volatilityfoundation/volatility3)
- Application for extracting PE images from Minidumps using [minidump](https://github.com/skelsec/minidump)
- Application for carving PE files from binary files
- Support for unpacking PE files from zip files (including password protected)
- Additional information in PE summary view
- Resource MIME type detection

### Changed
- Moved code for finding PE files in an IDB to the IDA runtime
- Switched to non-native file open dialogs as accept was slow on Windows
- Improved IDAPro in-memory PE handling
- Consolidated code in runtime
- Hide the output log view if it is empty

### Fixed

- Bug loading minidump files under IDA when no input file is present
- Added drop shadow to PE map labels to improve text readability
- Bug loading capstone disassembler
- Dumping without touching the IAT works properly now

## [1.0.29](/../../releases/tag/1.0.29) - 2020-10-05

### Fixed

- Fixed section MD5 VT search query

## [1.0.28](/../../releases/tag/1.0.28) - 2020-07-30

### Added

- [Rekall](http://www.rekall-forensic.com/) integration
- [.gitignore](.gitignore) file

### Changed

- Tidied runtime interface
- Renamed reconstructed imports section from .idata to .pe_tree

### Fixed

- Certificate parsing
- Improved PE dumping/import reconstruction
- pylint updates

## [1.0.27](/../../releases/tag/1.0.27) - 2020-05-20

### Added

- Added [LICENSE](LICENSE)

### Changed

- [setup.py](setup.py) now reads requirements from [requirements.txt](requirements.txt)
- Updated IDAPython installation documentation in [README.md](README.md)

### Fixed

- The PE no-overlay hash was not calculated correctly
- Ensure the correct tree item is removed from the view via right click -> remove

## [1.0.26](/../../releases/tag/1.0.26) - 2020-05-19

### Changed

- Removed *package_data* from [setup.py](setup.py)

### Added

- Added [MANIFEST.in](MANIFEST.in) to include package resources

## [1.0.25](/../../releases/tag/1.0.25) - 2020-05-19

### Changed

- Updated copyright
- Updated VirusTotal logo
- Updated about box information

## [1.0.24](/../../releases/tag/1.0.24) - 2020-05-19

### Added

- VirusTotal and CyberChef URLs moved to configuration file
- Improved developer documentation

### Changed

- Updated about box information

### Fixed

- Fixed fatal exception when starting pe_tree application with partial configuration
- Fixed several pylint warnings

## [1.0.23](/../../releases/tag/1.0.23) - 2020-05-15

### Changed

- Updated [setup.py](setup.py) to better support installing as either standalone application or IDA plugin

## [1.0.22](/../../releases/tag/1.0.22) - 2020-05-14

### Added

- Changelog markdown
- Contributing markdown

### Changed

- Updated [README.md](README.md)

### Fixed

- Fixed bug when loading as IDAPython plugin

## [1.0.21](/../../releases/tag/1.0.21) - 2020-05-12

### Added

- IAT reconstruction


pe_tree/ghidra.py
#
#   Copyright (c) 2021 BlackBerry Limited.  All Rights Reserved.
#
#   Licensed under the Apache License, Version 2.0 (the "License");
#   you may not use this file except in compliance with the License.
#   You may obtain a copy of the License at
#
#       http://www.apache.org/licenses/LICENSE-2.0
#
#   Unless required by applicable law or agreed to in writing, software
#   distributed under the License is distributed on an "AS IS" BASIS,
#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#   See the License for the specific language governing permissions and
#   limitations under the License.
#

"""PE Tree Ghidra application"""

# Standard imports
import os
import sys
from argparse import ArgumentParser, ArgumentDefaultsHelpFormatter

# Qt imports
from PyQt5 import QtCore, QtWidgets

# Ghidra imports
try:
    import ghidra_bridge
except ModuleNotFoundError:
    print("ghidra_bridge is required")
    exit()

import pe_tree.window
import pe_tree.runtime
import pe_tree.form
import pe_tree.info

class GhidraRuntime(pe_tree.runtime.Runtime):
    """Ghidra runtime callbacks"""
    def __init__(self, widget, args):
        # Load configuration
        self.config_file = os.path.join(self.get_temp_dir(), "pe_tree_ghidra.ini")
        super(GhidraRuntime, self).__init__(widget, args)

        # Initialise Ghidra bridge
        try:
            self.bridge = ghidra_bridge.GhidraBridge(connect_to_host=self.args.server, connect_to_port=self.args.port, namespace=globals(), response_timeout=20)
        except ConnectionRefusedError:
            print("Please run ghidra_bridge_server_background.py under Ghidra -> Window -> Script Manager")
            exit()

        state = getState()
        currentProgram = state.getCurrentProgram()

        self.address_factory = currentProgram.getAddressFactory()
        self.function_manager = currentProgram.getFunctionManager()
        self.data_type_manager = currentProgram.getDataTypeManager()
        self.listing = currentProgram.getListing()
        self.memory = currentProgram.getMemory()
        self.address_space = currentProgram.getAddressFactory().getDefaultAddressSpace()

    def jumpto(self, item, offset):
        """Jump to offset in Ghidra listing view"""
        try:
            self.ret = goTo(currentProgram.addressFactory.getAddress(hex(offset)))
        except:
            self.ret = False

        return self.ret

    @QtCore.pyqtSlot(object, object)
    def get_bytes(self, start, size):
        """Read bytes from memory"""
        array = self.bridge.remote_import("array")
        buffer = self.bridge.remote_eval("array.array('b', data)", array=array, data=b"\0" * size)

        read = self.memory.getBytes(self.address_space.getAddress(hex(start)), buffer)

        if read > 0:
            self.ret = bytearray(self.bridge.remote_eval("[ b & 0xff for b in data]", data=buffer))
        else:
            self.ret = b""

        return self.ret

    @QtCore.pyqtSlot(object)
    def get_name(self, offset):
        """Get name for offset from Ghidra"""
        self.ret = ""
        return self.ret

    @QtCore.pyqtSlot(object)
    def get_segment_name(self, offset):
        """Get segment name for offset from Ghidra"""
        self.ret = ""
        return self.ret

    @QtCore.pyqtSlot(object)
    def is_writable(self, offset):
        """Determine if memory address is writable"""
        self.ret = False
        return self.ret

    @QtCore.pyqtSlot(object)
    def resolve_address(self, offset):
        """Get module/symbol name for address"""
        symbol = ""
        module = ""

        ref = currentProgram.referenceManager.getReferencesFrom(self.address_factory.getAddress(hex(offset)))

        if ref:
            ref = ref[0].toString()

            # Should look something like.. "->MSVCRT.DLL::__setusermatherr"
            if ref.startswith("->"):
                # Remove ->
                ref = ref[2:]

                # Split into module/API
                elements = ref.split("::")

                module = elements[0]
                symbol = elements[1].split(" ")[0]

        self.ret = (module, symbol)
        return self.ret

    @QtCore.pyqtSlot(object)
    def get_label(self, offset):
        """Get label for offest from Ghidra"""
        self.ret = getSymbolAt(self.address_space.getAddress(hex(offset)))
        return self.ret

    @QtCore.pyqtSlot(object, object)
    def make_string(self, offset, size):
        """Convert range to byte string in Ghidra"""
        self.ret = self.listing.createData(self.address_space.getAddress(hex(offset)), ghidra.program.model.data.StringDataType, size)
        return self.ret

    @QtCore.pyqtSlot(object, object, str, str, bytes)
    def make_segment(self, offset, size, class_name="DATA", name="pe_map", data=None):
        """Create a new section in Ghidra - Not implemented"""
        self.ret = None
        return self.ret

    @QtCore.pyqtSlot(object)
    def make_qword(self, offset):
        """Create a qword at the given offset in the listing"""
        self.ret = self.listing.createData(self.address_space.getAddress(hex(offset)), ghidra.program.model.data.QWordDataType)
        return self.ret

    @QtCore.pyqtSlot(object)
    def make_dword(self, offset):
        """Create a dword at the given offset in the listing"""
        self.ret = self.listing.createData(self.address_space.getAddress(hex(offset)), ghidra.program.model.data.DWordDataType)
        return self.ret

    @QtCore.pyqtSlot(object)
    def make_word(self, offset):
        """Create a word at the given offset in the listing"""
        self.ret = self.listing.createData(self.address_space.getAddress(hex(offset)), ghidra.program.model.data.WordDataType)
        return self.ret

    @QtCore.pyqtSlot(object)
    def make_byte(self, offset, size=1):
        """Create a byte at the given offset in the listing"""
        self.ret = self.listing.createData(self.address_space.getAddress(hex(offset)), ghidra.program.model.data.ByteDataType)
        return self.ret

    @QtCore.pyqtSlot(object, str)
    def make_comment(self, offset, comment):
        """Create a comment at the given offset in the listing"""
        self.ret = self.listing.setComment(self.address_space.getAddress(hex(offset)), ghidra.program.model.listing.CodeUnit.REPEATABLE_COMMENT, comment)
        return self.ret

    def find_iat_ptrs_remote(self, image_base, size, get_word, namespace=globals()):
        """Find all likely IAT pointers
        
        Note! For performance reasons this code should be executed remotely under the Ghidra python interpreter using bridge.remoteify()

        """
        address_space = currentProgram.getAddressFactory().getDefaultAddressSpace()

        address_set = ghidra.program.model.address.AddressSet()
        address_set.addRange(address_space.getAddress(hex(image_base)), address_space.getAddress(hex(image_base + size)))

        opcode_iterator = currentProgram.getListing().getInstructions(address_set, True)

        # Iterate over all opcodes in the image
        iat_ptrs = []

        while opcode_iterator.hasNext():
            opcode = opcode_iterator.next()

            # Get the opcode mnemonic
            mnem = opcode.getMnemonicString().lower()

            # Does the opcode contain a memory address?
            ptr = 0

            if mnem in ["call", "push", "jmp"]:
                if opcode.getOperandType(0) & ghidra.program.model.lang.OperandType.ADDRESS:
                    # Get memory offset for branch instructions
                    ptr = opcode.getOpObjects(0)

            elif mnem in ["mov", "lea"]:
                if opcode.getOperandType(0) & ghidra.program.model.lang.OperandType.REGISTER and opcode.getOperandType(1) & ghidra.program.model.lang.OperandType.ADDRESS:
                    # Get memory offset for mov/lea instructions
                    ptr = opcode.getOpObjects(1)

            if ptr == 0:
                continue

            # Get the memory address value/offset
            try:
                ptr = int(ptr[0].getUnsignedValue())
            except:
                ptr = int(ptr[0].getOffset())

            # Does the instruction's memory address operand seem somewhat valid?!
            if ptr < 0x1000:
                continue

            # Read pointer from memory address
            try:
                # TODO: Ensure this is dynamic for x86/64!
                iat_offset = getInt(currentProgram.addressFactory.getAddress(hex(ptr)))
            except:
                iat_offset = 0

            if iat_offset == 0:
                continue

            # Ignore offset if it is in our image
            if iat_offset >= image_base and iat_offset < image_base + size:
                continue

            # Attempt to resolve reference (Note! do not use self.resolve_address, as this code is executed remotely ;)
            ref = currentProgram.referenceManager.getReferencesFrom(currentProgram.addressFactory.getAddress(hex(ptr)))

            if ref:
                ref = ref[0].toString()

                # Should look something like.. "->MSVCRT.DLL::__setusermatherr"
                if ref.startswith("->"):
                    # Remove ->
                    ref = ref[2:]

                    # Split into module/API
                    elements = ref.split("::")

                    module = elements[0]
                    api = elements[1].split(" ")[0]

                    # Add IAT offset, address to patch, module name and API name to list
                    iat_ptrs.append([iat_offset, ptr + opcode.getAddress().add(opcode.getLength() - 4).getOffset(), module, api])

        return iat_ptrs

    @QtCore.pyqtSlot(object, object, object, object)
    def find_iat_ptrs(self, pe, image_base, size, get_word):
        """Find all likely IAT pointers"""
        # Perform IAT search under Ghidra python interpreter 
        self.ret = self.bridge.remoteify(self.find_iat_ptrs_remote)(self, image_base, size, get_word)
        return self.ret

def main():
    """PE Tree Ghidra script entry-point"""
    # Check command line arguments
    parser = ArgumentParser(description="PE-Tree (Ghidra)", formatter_class=ArgumentDefaultsHelpFormatter)
    parser.add_argument("--server", help="Ghidra bridge server IP", default="127.0.0.1")
    parser.add_argument("--port", help="Ghidra bridge server port", default=4768, type=int)
    args = parser.parse_args()

    # Create PE Tree Qt application
    application = QtWidgets.QApplication(sys.argv)
    window = pe_tree.window.PETreeWindow(application, GhidraRuntime, args, open_file=False)

    # Try to locate the input file
    filename = currentProgram.getExecutablePath()

    if filename:
        # No input file found?
        if not os.path.isfile(filename):
            filename = os.path.basename(filename)

            if not os.path.isfile(filename):
                filename = runtime.ask_file(os.path.basename(filename), "Where is the input file?")

        if os.path.isfile(filename):
            # Map input file
            window.pe_tree_form.map_pe(image_base=currentProgram.getImageBase().getOffset(), filename=filename)

    sys.exit(application.exec_())

if __name__ == "__main__":
    main()